src/Data/Unrestricted/Internal/Dupable.hs

```diff
@@ -1,47 +1,72 @@
 {-# LANGUAGE DataKinds #-}
 {-# LANGUAGE GADTs #-}
 {-# LANGUAGE LinearTypes #-}
+{-# LANGUAGE PolyKinds #-}
 {-# LANGUAGE ScopedTypeVariables #-}
-{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE NoImplicitPrelude #-}
 {-# OPTIONS_HADDOCK hide #-}
 
 module Data.Unrestricted.Internal.Dupable
-  ( -- * Dupable
-    Dupable (..),
+  ( Dupable (..),
     dup,
     dup3,
+    dup4,
+    dup5,
+    dup6,
+    dup7,
   )
 where
 
-import Data.Type.Equality
+import Data.Replicator.Linear.Internal (Replicator (..))
+import qualified Data.Replicator.Linear.Internal as Replicator
+import Data.Replicator.Linear.Internal.ReplicationStream (ReplicationStream (..))
 import Data.Unrestricted.Internal.Consumable
-import Data.V.Linear.Internal.V (V)
-import qualified Data.V.Linear.Internal.V as V
-import GHC.TypeLits
+import Prelude.Linear.Internal
 
--- | The laws of @Dupable@ are dual to those of 'Monoid':
+-- | The laws of 'Dupable' are dual to those of 'Monoid':
```

Comment:
We are missing laws for `dupR`.

Let's see. We need

```haskell
dup2 = Replicator.elim (,) . dupR
```

Correspondingly for `consume`

```haskell
consume = Replicator.consume . dupR
```

We could state this one another way, like below. Maybe it's better, actually.

```haskell
consume = Replicator.elim () . dupR
```

And we need laws for the comonadic structure

```haskell
* Replicator.extract . dupR = id
* Replicator.duplicate . dupR = (Replicator.map dupR) . dupR
```

These laws make the `consume`/`dup2` laws redundant. But we should leave them in as they are probably easier to understand.
```

Reply:
> And we need laws for the comonadic structure
    * Replicator.extract . dupR = id
    * Replicator.duplicate . dupR = (Replicator.map dupR) . dupR

I don't think that we need to add these laws. `ReplicationStream` and `Replicator` constructors are not exposed, so the user cannot implement something where they would fail. At the moment, `dupR` can only be defined by the user via `dup2` (where `give == id`), or via `AsMovable` (where `give == unur`). Also, `Replicator.{extract,duplicate,map}` are black boxes for the user, so I don't see how he could check that the laws hold. Finally, `Replicator.duplicate` is `dupR`.

Maybe we could write these laws in the internal documentation of `Replicator` (or `ReplicationStream`) though.

Reply:
Here is a way to write a faulty `Dupable` instance:

```haskell
instance Dupable Bool where
  dupR b = b `lseq` pure True
```

Reply:
> Here is a way to write a faulty `Dupable` instance:
> 
> ```haskell
> instance Dupable Bool where
>   dupR b = b `lseq` pure True
> ```

But in this case, either `dup2` is deduced from `dupR` and then, the `dup2` neutrality rule is violated, or `dup2` is implemented in a law-abiding way, and then, the `dup2`-`dupR` coherence rule is violated. My point is, I don't think that the user can implement something that would respect the first four rules, but would violate one of the comonadic-related laws.

Reply:
Ah, I see what you mean.

I can't think of a way to define `dupR` such that `dup2` is correct, and associative and such, yet these laws are violated. It might exist in principle though (whereas if the `dupR` laws are correct, then `dup2`, being uniquely determined by `dupR`, will clearly have the laws validated (well, I say “clearly” I don't really anticipate that the proof will be easy actually)).

That being said, it doesn't hurt writing these laws as a guideline for someone who'd only write a `dupR` instance, rather than having them work backward from `dup2`. The laws are definitely redundant, and we can say it if you feel more comfortable. But I still think that we should have them.